discovery/file/file.go

```diff
@@ -233,19 +317,28 @@ func (d *Discovery) refresh(ctx context.Context, ch chan<- []*config.TargetGroup
 	d.watchFiles()
 }
 
-// fileSource returns a source ID for the i-th target group in the file.
-func fileSource(filename string, i int) string {
-	return fmt.Sprintf("%s:%d", filename, i)
-}
-
 // readFile reads a JSON or YAML list of targets groups from the file, depending on its
 // file extension. It returns full configuration target groups.
-func readFile(filename string) ([]*config.TargetGroup, error) {
-	content, err := ioutil.ReadFile(filename)
+func (d *Discovery) readFile(filename string) ([]*config.TargetGroup, error) {
+	fd, err := os.Open(filename)
 	if err != nil {
 		return nil, err
 	}
 
+	content, err := ioutil.ReadAll(fd)
+	if err != nil {
+		return nil, err
+	}
+
+	// We use a mutex when setting timestamp, so don't block the rest of readFile.
```

Comment:
There's no need for a goroutine her, a single syscall isn't that expensive.

Reply:
This was more for the writing of the timestamp to `Discovery`'s map, I thought if there were a lot of files we wouldn't want to block the reading of each one on getting and storing the timestamp. I'll remove it if you still think it's unnecessary.

Reply:
Updating a map is cheaper than creating a goroutine.